ktor-client/ktor-client-features/ktor-client-auth/common/src/io/ktor/client/features/auth/providers/BearerAuthProvider.kt

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright 2014-2020 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.client.features.auth.providers
+
+import io.ktor.client.call.*
+import io.ktor.client.features.auth.*
+import io.ktor.client.request.*
+import io.ktor.http.*
+import io.ktor.http.auth.*
+import kotlinx.atomicfu.*
+import kotlinx.coroutines.*
+
+/**
+ * Add [BasicAuthProvider] to client [Auth] providers.
+ */
+public fun Auth.bearer(block: BearerAuthConfig.() -> Unit) {
+    with(BearerAuthConfig().apply(block)) {
+        providers.add(BearerAuthProvider(_refreshTokens, _loadTokens, true, realm))
+    }
+}
+
+public data class BearerTokens(
+    val accessToken: String,
+    val refreshToken: String
+)
+
+/**
+ * [DigestAuthProvider] configuration.
+ */
+public class BearerAuthConfig {
+    internal var _refreshTokens: suspend (call: HttpClientCall) -> BearerTokens? = { null }
+    internal var _loadTokens: suspend () -> BearerTokens? = { null }
+
+    public var realm: String? = null
+
+    public fun refreshTokens(block: suspend (call: HttpClientCall) -> BearerTokens?) {
+        _refreshTokens = block
+    }
+
+    public fun loadTokens(block: suspend () -> BearerTokens?) {
+        _loadTokens = block
+    }
+}
+
+/**
+ * Client digest [AuthProvider].
+ */
+public class BearerAuthProvider(
+    public val refreshTokens: suspend (call: HttpClientCall) -> BearerTokens?,
+    public val loadTokens: suspend () -> BearerTokens?,
+    override val sendWithoutRequest: Boolean = true,
+    private val realm: String?
+) : AuthProvider {
+    private val cachedBearerTokens: AtomicRef<CompletableDeferred<BearerTokens?>> =
+        atomic(CompletableDeferred<BearerTokens?>().apply {
+            complete(null)
+        })
+
+    /**
+     * Check if current provider is applicable to the request.
+     */
+    override fun isApplicable(auth: HttpAuthHeader): Boolean {
+        if (auth.authScheme != AuthScheme.Bearer) return false
```

Comment:
I thought we always wrap `if` body into `{}`, but I like this codestyle more, so it's OK for me.